tests/main/snap-quota-cpu/task.yaml

```diff
@@ -0,0 +1,129 @@
+summary: Functional test for cpu quota-related snap commands.
+
+details: |
+  Functional test for snap cpu quota group commands ensuring that they are 
+  effective in practice.
+
+# these systems do not support quota groups due to their old systemd versions,
+# we do have a check to do for these systems, but that is just to ensure that
+# we don't allow using quota groups, and that check is done in the snap-quota
+# spread instead
+systems:
+  - -ubuntu-14.04-*
+  - -centos-7-*
+  - -amazon-linux-2-*
+  - -ubuntu-16.04-*
+  - -ubuntu-core-16-*
+
+prepare: |
+  snap install test-snapd-stressd --edge --devmode
+  snap set system experimental.quota-groups=true
+  tests.cleanup defer snap unset system experimental.quota-groups
+
+execute: |
+  echo "Starting service and verifying cpu usage"
+  snap start test-snapd-stressd.stress-sc
+
+  # helper function that has retry logic to allow for cpu usage to stabilize
+  function ensure_cpu_usage_higher_than {
+    n=0
+    until [ "$n" -ge 5 ]
+    do
+      usage=$(top -b -n 2 -d 0.2 | head | grep "$1" | awk '{print $9*100}')
+      if [ "$usage" -gt "$2" ]; then
+        return 0
+      fi
+      n=$((n+1))
+      sleep 1
+    done
+
+    echo "Expected cpu usage to be minimum $2%, got $usage"
+    exit 1
+  }
+  
+  echo "Verifying cpu usage, expecting atleast above 60% as we test against 50%"
+  ensure_cpu_usage_higher_than "stress" 6000
+
+  echo "Create a group with a snap in it"
+  snap set-quota group-one --cpu=2x25% test-snapd-stressd
+
+  echo "The systemd slice should be active now"
+  sliceName="snap.$(systemd-escape --path group-one).slice"
+  systemctl show --property=ActiveState "$sliceName" | MATCH "ActiveState=active"
+
+  echo "The service should also still be active"
+  snap services test-snapd-stressd.stress-sc | MATCH "test-snapd-stressd.stress-sc\s+disabled\s+active"
+
+  # systemd/kernel have three different locations for the cgroup pids depending
+  # on version
+  echo "The systemd slice should have two processes (all from stress snap) in it now"
+  cgroupsV1OldSystemdProcsFile="/sys/fs/cgroup/cpuacct/$sliceName/snap.test-snapd-stressd.stress-sc.service/cgroup.procs"
+  cgroupsV1ProcsFile="/sys/fs/cgroup/cpuacct/$sliceName/cgroup.procs"
+  cgroupsV2ProcsFile="/sys/fs/cgroup/$sliceName/snap.test-snapd-stressd.stress-sc.service/cgroup.procs"
+  if [ -e "$cgroupsV2ProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV2ProcsFile"
+  elif [ -e "$cgroupsV1OldSystemdProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV1OldSystemdProcsFile"
+  elif [ -e "$cgroupsV1ProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV1ProcsFile"
+  else
+      echo "cannot detect cgroup procs file"
+      exit 1
+  fi
+
+  #shellcheck disable=SC2016
+  retry --wait 1 -n 100 --env cgroupProcsFile="$cgroupProcsFile" sh -x -c 'test "$(wc -l < $cgroupProcsFile)" = 2'
```

Comment:
(nitpick) why wait 100s for this? just curious, maybe/probably fine, just feels long. (nitpick^9) --wait=1 is the default iirc ;)

Reply:
It does say `--wait 1`? just with a 100 iterations, so it hopefully doesn't wait 100s (unless my eyes are deceiving me!)